### ISSUE TITLE
Feature Request: Ability to install dependencies at a file level.

### ISSUE THREAD

**Issue by reggi** [npm/npm#9456]
Would love to run a command like  `npm install depsfrom ./index.js`, and have the script check the file for which dependencies it needs using something like [acorn-umd](https://github.com/megawac/acorn-umd) which crawls the AST of the javascript document looking for module definitions, then install the deps. 

I'd love to make this if it doesn't exist already. Wondering `npm`'s stance on this functionality in the future. Thoughts? Feel free to close ASAP.

**Comment by KenanY** [comment]
@reggi Reminds me of [lint-deps](https://github.com/jonschlinkert/lint-deps) :)

**Comment by othiym23** [comment]
This sounds like a use case for `npm clone`, a hypothetical feature that @seldo has requested which will download a package tarball, unpack it into the current directory, and then peace out. A request from the CTO isn't an irresistible force, but it's not far off, so we'll probably build something like this, although not soon. If you need it for your own purposes, a PR would be welcome, @reggi.

Most of the plumbing is already there, in the form of `npm cache add` -- there just needs to be a further task that uses `lib/utils/tar.js` to unpack the cached tarball into the `npm.prefix` for the current process. I'd be happy to work with you to get this in. If you want to make it possible to extract an individual file, that shouldn't be too hard either, but the default behavior of `npm clone <package spec>` should be to just unpack the whole thing.

**Comment by othiym23** [comment]
I went ahead and [added this to the roadmap](https://github.com/npm/npm/wiki/Roadmap#front-end-asset-installer), as a specific feature request tied to @seldo's request. @reggi, let me know if my understanding of what you want differs from what @seldo has described he wants, and we'll see if we can come up with one feature that serves both needs.

**Comment by reggi** [comment]
@KenanY Thanks a bunch! lint-deps works pretty much as I would like it gets the deps that are needed then installs them. The only problem I have with it is that it doesn't just install them it asks you for input / what you want to do. 

@othiym23 I think you've nailed it. Here's a walkthrough.

```
.
└── index.js
```

Where `index.js` contains:

```js
var _ = require('lodash')
var Promise = require('bluebird')
```

*  `npm <COMMAND> ./index.js` would install `lodash`  & `bluebird`
*  `npm <COMMAND> ./index.js --save` would install & save `lodash`  & `bluebird`
*  `npm <COMMAND> ./index.js --save-dev` would install & save as dev dep  `lodash`  & `bluebird`

**Comment by reggi** [comment]
Ideally this can also recursively get deps for files / a directory as well.

I'd also want the ability to not install but just report on what deps are missing. 

Where:

```
npm <COMMAND> index.js --report
```

Will provide `lodash, bluebird missing` and exit with non-zero status code.

And if all is well it would exit with zero.

**Comment by othiym23** [comment]
You seem to have something very different in mind from what I thought Ibwas describing -- I'm only described by a command that would allow you to unpack a package tarball or one or more files in that tarball, explicitly specified. You're describing something that would unpack a file _and parse it_, selectively unpacking _or installing_ anything that file requires. That is _much_ more work to build, and I think isn't something that most npm users would need or use. It would be a cool thing to build an external tool to do, but isn't something I think makes sense for npm itself. The simpler thing I described originally is something that I'm interestedly in seeing in npm, and could be tied to something that post processes in the way you're describing via a run script.

**Comment by reggi** [comment]
@othiym23 Whoops sorry about that. I'll talk to the creator of lint-deps about getting some of my features into it. Thanks for clarifying and sorry for the confusion!

**Comment by watilde** [comment]
This feature request seems actually interesting.

@othiym23 Could I ask you the final sketch of this feature is like the following commands?
+ npm clone `<tarball file>`
+ npm clone `<folder>`
+ npm clone `<tarball url>`
+ npm clone `<git url>`
+ npm clone `<name>@<version>`

e.g.
```
$ npm clone http://registry.npmjs.org/text-table/-/text-table-0.2.0.tgz
success!

$ ls
package.json text-table node_modules
```

**Comment by othiym23** [comment]
@watilde Sorry for not getting back to you on this. I've been busy, but that's not an excuse, because I do want people to know about both `npm clone`, and the _new_ command I think this resembles, which I've called `npm sync`, but probably need to find a better name for, because it's pretty overloaded. The thing you're sketching out is yet another interesting tool that would be good to talk about, but looks pretty different from what either I or @reggi are thinking of (start another discussion?). Names are hard!

@reggi Now that I have learned to read, and know what you were actually suggesting, I think this is a good idea. I also think, per @KenanY, this is something that is best prototyped outside of the CLI, because there's a fair amount of new functionality to play around with when you're introducing static analysis of dependencies and it would be good to get some hands-on experience with it before integrating into a tool operating at npm's scale. So I'm going to close this feature request (there are way too many of them open for the team to get to them all / for them to be useful as a tracking tool).

However! If something like `lint-deps` does get to a useful place, and people are interested in integrating it into the CLI, I think this is something that would be useful in both reading dependencies from source _and_ in linting packages prepublish (I really want `npm lint` to be a thing someday, or maybe a new tool altogether). So a PR adding something like this to those would be most welcome.
